drivers/serial/uart_bflb.c

```diff
@@ -0,0 +1,639 @@
+/*
+ * Copyright (c) 2021-2024 ATL Electronics
+ * Copyright (c) 2024 MASSDRIVER EI (massdriver.space)
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT bflb_bl_uart
+
+/**
+ * @brief UART driver for Bouffalo Lab MCU family
+ */
+
+#include <zephyr/drivers/uart.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/arch/common/sys_io.h>
+#include <zephyr/drivers/timer/system_timer.h>
+#include <zephyr/pm/policy.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/irq.h>
+
+#include <glb_reg.h>
+#include <hbn_reg.h>
+#include <pds_reg.h>
+#include <bouffalolab/common/uart_reg.h>
+#include <bouffalolab/common/bflb_uart.h>
+
+/* clang-format off */
+
+struct bflb_config {
+	const struct pinctrl_dev_config *pincfg;
+	uint32_t baudrate;
+	uint8_t direction;
+	uint8_t data_bits;
+	uint8_t stop_bits;
+	uint8_t parity;
+	uint8_t bit_order;
+	uint8_t flow_ctrl;
+	uint8_t tx_fifo_threshold;
+	uint8_t rx_fifo_threshold;
+	uint32_t base_reg;
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_config_func_t irq_config_func;
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+struct bflb_data {
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_callback_user_data_t user_cb;
+	void *user_data;
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+static void uart_bflb_enabled(const struct device *dev, uint32_t enable)
+{
+	const struct bflb_config *cfg = dev->config;
+	uint32_t rxt = 0;
+	uint32_t txt = 0;
+
+	if (enable) {
+		enable = 1;
+	}
+
+	txt = sys_read32(cfg->base_reg + UART_UTX_CONFIG_OFFSET);
+	txt = (txt & ~UART_CR_UTX_EN) | enable;
+	rxt = sys_read32(cfg->base_reg + UART_URX_CONFIG_OFFSET);
+	rxt = (rxt & ~UART_CR_URX_EN) | enable;
+	sys_write32(rxt, cfg->base_reg + UART_URX_CONFIG_OFFSET);
+	sys_write32(txt, cfg->base_reg + UART_UTX_CONFIG_OFFSET);
+}
+
+static uint32_t uart_bflb_get_crystal_frequency(void)
+{
+	uint32_t tmp;
+
+	/* get clkpll_sdmin */
+	tmp = sys_read32(PDS_BASE + PDS_CLKPLL_SDM_OFFSET);
+	tmp = (tmp & PDS_CLKPLL_SDMIN_MSK) >> PDS_CLKPLL_SDMIN_POS;
+
+	/*
+	 * The 'Sigma Delta Modulator Input' scales the PLL output relative to the
+	 * input frequency (eg value is x2, output frequency doubles), it's a
+	 * direct indicator of the XCLK (crystal or 32M osc) frequency to go in the
+	 * PLL.
+	 *
+	 * This value can be read from efuse or from the firmware header which will
+	 * be implemented in the future.
+	 */
+	switch (tmp) {
+	case 0x500000:
+		return MHZ(24);
+	case 0x3C0000:
+		return MHZ(32);
+	case 0x320000:
+		return MHZ(38.4);
+	case 0x300000:
+		return MHZ(40);
+	case 0x49D39D:
+		return MHZ(26);
+	default:
+		return MHZ(32);
+	}
+}
+
+static uint32_t uart_bflb_get_PLL_frequency(void)
```

Comment:
Function names are typically lowercase but it might be an exception for external SDKs, as for which obviously not possible to use anything else than what the vendor provides...
Here it is a Zephyr driver though, so might be possible to change.
I agree that `pll` looks a bit less like "Phased Locked Loop" and a bit more "pelll"...

Reply:
I fixed to use lowercase. Everything related to clocks will be handled in future PRs.